substrate/frame/revive/src/precompiles.rs

```diff
@@ -0,0 +1,587 @@
+	///
+	/// Those are the errors that are commonly caught by Solidity try-catch blocks. Encodes
+	/// a string onto the output buffer.
+	Revert(Revert),
+	/// An error generated by Solidity itself.
+	///
+	/// Encodes an error code into the output buffer.
+	Panic(PanicKind),
+	/// Don't encode anything into the output buffer. Just trap.
+	///
+	/// Commonly used for out of gas or other resource errors.
+	Error(ExecError),
+}
+
+impl From<DispatchError> for Error {
+	fn from(error: DispatchError) -> Self {
+		Self::Error(error.into())
+	}
+}
+
+impl<T: Config> From<CrateError<T>> for Error {
+	fn from(error: CrateError<T>) -> Self {
+		Self::Error(DispatchError::from(error).into())
+	}
+}
+
+/// Type that can be implemented in other crates to extend the list of pre-compiles.
+///
+/// Only implement exacly one function. Either `call` or `call_with_info`.
+///
+/// # Warning
+///
+/// Pre-compiles are unmetered code. Hence they have to charge an appropriate amount of weight
+/// themselves. Generally, their first line of code should be a call to
+/// `env.gas_meter_mut().charge()`. For that you need to implement [`Token`] on a type of your
+/// chosing.
+pub trait Precompile {
+	/// Your runtime.
+	type T: Config;
+	/// The Solidity ABI definition of this pre-compile.
+	///
+	/// Use the [`self::alloy::sol`] macro to define your interface using Solidity syntax.
+	/// The input the caller passes to the pre-compile will be validated and parsed
+	/// according to this interface.
+	///
+	/// Please note that the return value is not validated and it is the pre-compiles
+	/// duty to return the abi encoded bytes conformant with the interface here.
+	type Interface: SolInterface;
+	/// Defines at which addresses this pre-compile exists.
+	const MATCHER: AddressMatcher;
+	/// Defines whether this pre-compile needs a contract info data structure in storage.
+	///
+	/// Enabling it unlocks more APIs for the pre-compile to use. Only pre-compiles with a
+	/// fixed matcher can set this to true. This is enforced at compile time. Reason is that
+	/// contract info is per address and not per pre-compile. Too many contract info structures
+	/// and accounts would be created otherwise.
+	///
+	/// # When set to **true**
+	///
+	/// - An account will be created at the pre-compiles address when it is called for the first
+	///   time. The ed is minted.
+	/// - Contract info data structure will be created in storage on first call.
+	/// - Only `call_with_info` should be implemented. `call` is never called.
+	///
+	/// # When set to **false**
+	///
+	/// - No account or any other state will be created for the address.
+	/// - Only `call` should be implemented. `call_with_info` is never called.
+	///
+	/// # What to use
+	///
+	/// Should be set to false if the additional functionality is not needed. A pre-compile with
+	/// contract info will incur both a storage read and write to its contract metadata when called.
```

Comment:
why a write?

Reply:
At the end of the execution of the pre-compile we need to write back the changes.